File: 391-perfect-rectangle/solution.py
class Solution:
    def isRectangleCover(self, rectangles: List[List[int]]) -> bool:
        X1, Y1 = float('inf'), float('inf')
        X2, Y2 = -float('inf'), -float('inf')
        
        points = set()
        actual_area = 0
        for x1, y1, x2, y2 in rectangles:
            # 计算完美矩形的理论顶点坐标
            X1, Y1 = min(X1, x1), min(Y1, y1)
            X2, Y2 = max(X2, x2), max(Y2, y2)
            # 累加小矩形的面积
            actual_area += (x2 - x1) * (y2 - y1)
            # 记录最终形成的图形中的顶点
            p1, p2 = (x1, y1), (x1, y2)
            p3, p4 = (x2, y1), (x2, y2)
            for p in [p1, p2, p3, p4]:
                if p in points: points.remove(p)
                else:           points.add(p)
        # 判断面积是否相同
        expected_area = (X2 - X1) * (Y2 - Y1)
        if actual_area != expected_area:
            return False
        # 判断最终留下的顶点个数是否为 4
        if len(points) != 4:       return False
        # 判断留下的 4 个顶点是否是完美矩形的顶点
        if (X1, Y1) not in points: return False
        if (X1, Y2) not in points: return False
        if (X2, Y1) not in points: return False
        if (X2, Y2) not in points: return False
        # 面积和顶点都对应，说明矩形符合题意
        return True

File: 391-perfect-rectangle/solution.cpp
class Solution {
public:
    bool isRectangleCover(vector<vector<int>>& rectangles) {
        int X1 = INT_MAX, Y1 = INT_MAX;
        int X2 = INT_MIN, Y2 = INT_MIN;
        using point = pair<int, int>;
        set<point> points;
        long recAreas = 0;
        for (auto & rec: rectangles) {
            int x1 = rec[0], y1 = rec[1];
            int x2 = rec[2], y2 = rec[3];
            X1 = min(x1, X1), Y1 = min(y1, Y1);
            X2 = max(x2, X2), Y2 = max(y2, Y2);
            recAreas += ((long)x2 - (long)x1) * ((long)y2 - (long)y1);
            
            for (point & p: vector<point>({{x1, y1}, {x1, y2}, {x2, y1}, {x2, y2}})) {
                if (points.find(p) != points.end()) {
                    points.erase(p);
                } else {
                    points.emplace(p);
                }
            }
        }

        if(points.size() != 4) return false;

        if (points.find({X1, Y1}) == points.end()) return false;
        if (points.find({X1, Y2}) == points.end()) return false;
        if (points.find({X2, Y1}) == points.end()) return false;
        if (points.find({X2, Y2}) == points.end()) return false;

        if (((long)X2 - (long)X1) * ((long)Y2 - (long)Y1) != recAreas) {
            return false;
        }

        return true;
    }
};